### PR TITLE
Add PRM training with hard estimation

### PR DESCRIPTION
**Pull request by zhuzilin**
This is an attempt to add process reward model (PRM) training. Currently I'm focusing on training with hard estimation loss in [Math-Shepherd](https://arxiv.org/abs/2312.08935) with its open sourced data [peiyi9979/Math-Shepherd](https://huggingface.co/datasets/peiyi9979/Math-Shepherd) and [mistral-7B](https://huggingface.co/mistralai/Mistral-7B-v0.1) (the same as [peiyi9979/math-shepherd-mistral-7b-prm](https://huggingface.co/peiyi9979/math-shepherd-mistral-7b-prm) ).

The current testing script is:
```bash
deepspeed --module openrlhf.cli.train_prm \
   --save_path ./checkpoint/mistal-7b-prm \
   --save_steps 500 \
   --logging_steps 1 \
   --eval_steps 100 \
   --train_batch_size 256 \
   --micro_train_batch_size 8 \
   --pretrain mistralai/Mistral-7B-v0.1  \
   --bf16 \
   --max_epochs 1 \
   --max_len 8192 \
   --zero_stage 3 \
   --learning_rate 1e-6 \
   --dataset peiyi9979/Math-Shepherd \
   --input_key input \
   --label_key label \
   --flash_attn \
   --load_checkpoint \
   --gradient_checkpointing \
   --packing_samples \
   --wandb_group prm \
   --placeholder_token "ки" \
   --reward_tokens "+" "-"
```

Note that the hyperparams are set casually, without reference (except the epoch and lr are the same as the paper). The resulting wandb is: https://wandb.ai/zhuzilin/openrlhf_train_prm/workspace

### CODE REVIEW

File: openrlhf/models/loss.py [hunk 1]
```diff
@@ -254,3 +254,34 @@ def forward(self, logits: torch.Tensor, teacher_logits: torch.Tensor, label: tor
...
+
+    def forward(self, inputs: torch.Tensor, logits: torch.Tensor, labels: torch.Tensor, *, return_acc: bool = False):
+        placeholder_mask = inputs == self.placeholder_token_id
+        logits = logits[placeholder_mask]
+        labels = labels[placeholder_mask]
+        if self.reward_token_ids is not None:
+            logits = logits[..., self.reward_token_ids]
+            logits = logits.softmax(dim=-1)
+            # this is bad....
+            for i, token in enumerate(self.reward_token_ids):
```
Comment (catqaq) [inline openrlhf/models/loss.py:279]: better to use matrix op instead of for loop

**Comment by catqaq** [comment]
The current implementation seems to be incompatible with qwen. only for mistral-7B?
